(PUP-6295) Add corrective_change field to report and report events

kbarber opened the pull request:
The corrective_change field in events marks an event that was presumed to be a
manual out-of-band change from Puppet, that Puppet then corrected. In the
context of a noop or non-applying change, the same is true except that the
change was not corrected but is intended to be corrected once a real run is
applied.

This patch introduces the work necessary to calculate this change, introducing a
transaction store that stores the last known best applied value so we can
calculate that a change was indeed unexpected.

For each report, we also mark the resource that had a corrective_change plus
mark an overall report of this also so this information can be stored in
backends like PuppetDB for quick searching.

Signed-off-by: Ken Barber <ken@bob.sh>

kbarber commented:
Replaces #5108

puppetcla commented:
CLA signed by all contributors.

adrienthebo commented on lib/puppet/transaction/event.rb:
```diff
@@ -94,6 +98,28 @@ def to_yaml_properties
+  def calculate_corrective_change(old_system_value)
+    # Only idempotent properties, and cases where we have an old system_value
+    # are corrective_changes.
+    @corrective_change = if @property_instance.idempotent? &&
```
This combines an `if` expression that has side effects and uses the return value, which is confusing. I would prefer something like this:

```ruby

@corrective_change = @property_instance.idempotent? && !@property_instance.sensitive && !old_system_value.nil?

if @corrective change
  # ...
else
  false
end
```

kbarber replied:
Okay, but @corrective_change doesn't equal the first part of the `if` clause, its equal to the insync? result further down. I can rework this though, to not use the return value of the `if` easily enough.

adrienthebo commented on lib/puppet/transaction/persistence.rb:
```diff
@@ -0,0 +1,81 @@
+    end
+
+    result = nil
+    Puppet::Util.benchmark(:debug, "Loaded transaction") do
```
`"Loaded transaction"` - what transaction?

kbarber replied:
Should be loaded transaction state file.

adrienthebo commented on lib/puppet/transaction/persistence.rb:
```diff
@@ -0,0 +1,81 @@
+    end
+
+    unless result.is_a?(Hash)
+      Puppet.err "State got corrupted"
```
I've seen this error message pop up when using Puppet and I've found it really confusing; can we indicate that the persistence file itself was corrupted?

kbarber replied:
Can do.

adrienthebo commented on lib/puppet/transaction/report.rb:
```diff
@@ -386,6 +395,16 @@ def self.default_format
   private
 
+  # Mark the report as corrective, if there are any resource_status marked corrective.
+  def calculate_report_corrective_change
```
:+1:

kbarber replied:
Cool bananas.

Iristyle replied:
Can be shortened to `resource_statuses.any?`

kbarber replied:
@Iristyle good thinking, done.

adrienthebo commented on lib/puppet/transaction/resource_harness.rb:
```diff
@@ -92,6 +93,21 @@ def perform_changes(resource, context)
+    persist_system_values(resource, context)
+  end
+
+  def persist_system_values(resource, context)
```
A system_value is actual system state after resource application?

kbarber replied:
Yes, thats basically it.

kbarber replied:
I forgot a docstring here, I'll add it.

kbarber commented:
@adrienthebo amended based on your comments.

adrienthebo commented on lib/puppet/transaction/report.rb:
```diff
@@ -1,5 +1,6 @@
 require 'puppet'
 require 'puppet/indirector'
+require 'puppet/transaction/persistence'
```
Small note: I don't think we actually use the `P::T::Persistence` in this file.

kbarber replied:
Cool fixed.

adrienthebo commented on lib/puppet/resource/catalog.rb:
```diff
@@ -214,6 +219,7 @@ def alias(resource, key)
   # @api public
   def apply(options = {})
     Puppet::Util::Storage.load if host_config?
+    persistence.load if host_config?
```
I see why we're loading the persistence file here, but I don't like it. We're adding in the concept of persistence to the catalog which doesn't need to know about it; it just happens that `Puppet::Resource::Catalog#apply` happens to manage the lifetime of the transaction (and knows that we're a host catalog) and the `Storage` class was loaded and saved here, but really neither of these are related to the catalog. I would rather have this moved to `Puppet::Transaction#evaluate` to manage the persistence object lifetime; at that point `Catalog#host_config?` should still be accessible and we won't have to have the persistence file managed so far way from where it's used.

kbarber replied:
No worries, fixed.

adrienthebo commented on lib/puppet/transaction/event.rb:
```diff
@@ -94,6 +98,30 @@ def to_yaml_properties
+
+      # Here we duplicate the property to avoid side-effects, and modify its should value
+      # so we can use insync?
+      test_property = @property_instance.dup
```
This scares the crap out of me. The RAL isn't intended to duplicate properties like this, and while this is possible it's working against how things are built. My biggest concern is that I can't predict how this will behave but I do suspect that this might not be safe.

/cc @branan

kbarber replied:
So its a squirrely one. I've incidentally now pushed this logic into the property so we can see it on its own. Let me point at that code to discuss this further.

kbarber commented on lib/puppet/property.rb:
```diff
@@ -346,6 +362,21 @@ def insync?(is)
+  # @return [Boolean] whether or not the values are in sync or not
+  # @api private
+  def insync_values?(orig_should, orig_is)
+    test_property = self.dup
```
@adrienthebo @branan so here is the code in question. I've moved this into `property.rb` from `event.rb`, because its weird and weirdness like this probably belongs closer to the problem.

So ... following on from the previous PR comment ... this methodology provides results, but as you say you are concerned with the side-effects of a dup.

Let me begin by outlining our limitations. Today the insync? comparison is my best method for proving things are 'equal' or 'insync' after the fact. Without it, ensure properties won't match for example (say `present => absent` on a file fails because its really `file => absent`).

However the API of `#insync?` dictates you must set the `@should` instance variable (in this case via a `#should=` method) to do so. Unfortunately, `#insync?` is an abstract method often overwritten by custom types, and so an overridden method expects this as its API to work with. To point, we're limited to what we can change because it will break custom extensions.

As far as other ideas and implementations to achieve similar results ... I've incidentally tried another methodology, whereby I backed up the old value, do an insync? and restore the value, this at the time produced odd results that I didn't understand (as in, it was marking things out of sync when they were not). Unless you have a better idea, I can resurrect that methodology ^^ for investigation.

branan replied:
I'm also not super thrilled with this, but I agree with your reasoning, Ken :(

If we were dup'ing a property in a more complex situation I'd be a lot more scared, but given that this is tiny function and the lifetime of the dup'd property is easy to understand, I think it's OK-ish. Just don't let it escape into the wild where somebody might call #sync on it, or we're gonna have a situation

![](https://media.giphy.com/media/eLpoGELHst13a/giphy.gif)

kbarber replied:
Lol. Its a pita. I can always right a dpittman-style essay on why its bad in the comments :-).

I'll try to do the other methodology, see how far I get, at least try to understand why it doesn't work. Having it contained like you state feels better than doing it from-a-distance at least.

adrienthebo replied:
I would be happy for a dpittman essay, so it's clear to anyone that sees this that we don't want this behavior to replicate.

kbarber replied:
if its contained, and the API is good enough, we can fix the implementation presumably.

+1 on big fat comment, if we're up for a type refactor in the future - people will want reasons.

Also, unit tests. I was waiting for stability of structure.

Iristyle commented on lib/puppet/type/notify.rb:
```diff
@@ -6,7 +6,7 @@ module Puppet
     @doc = "Sends an arbitrary message to the agent run-time log."
 
-    newproperty(:message) do
+    newproperty(:message, :idempotent => false) do
```
Is it better to modify the underlying property like you have you done here, or would it be less disruptive to provide a filter / whitelist somewhere later?

kbarber replied:
This has been debated both ways, @adrienthebo had liked this methodology I think. Allowing someone to mark something via the type interface does have the benefit of re-usability, and in some ways its neater because of that, but we don't know who needs it yet TBH, so this could be YAGNI. One could argue things in the wild might have known idempotency issues, but a flag to hide that might not be what is desired I guess.

There has been some debating of also filtering exec, but I'm unsure if the same implementation applies in this case, since we wouldn't filter because its inherently non-idempotent but for other boring reasons I can get into if you're curious.

kbarber replied:
@Iristyle have you talked me out of this implementation? This is public API, and more knobs without reason suck. @adrienthebo, would a whitelist style implementation suck? Can we do something like this, and make it private I wonder? Being able to mark idempotency issues for other custom types really just fell out of this effort, its not essential to success (accidental feature?), at least I'm not entirely sure it is. We can always add it in, when its called for?

adrienthebo replied:
If we do a whitelist then that's probably going to have to live in core; with an idempotent property that can be done on a per-custom type basis which is nice. Also, this implementation works (it works, right?) and we're under pressure to ship, so I would rather just take this as-is and roll with it. If this is YAGNI then it's very small YAGNI with limited scope and minimal impact.

Iristyle replied:
You raise a good point @adrienthebo - without a mechanism for extending the whitelist, type and provider builders can't opt into the "non-idempotent" property behavior. Hmm...

Iristyle replied:
Thinking about this a little more... what happens when modules you use trigger errant `corrective_change` events?  And the module is not being updated by the maintainer, etc.

kbarber replied:
So bugs that cause idempotency issues will show as corrective_change's today, this is a known limitation. Also some DSL code can do this also, like services with incorrect status handling and no status => overrides.

The future, would hopefully be able to bring detecting differences between bugs and true corrective_changes, but we were trying to avoid boiling the ocean for now.

If you're interested I have some points on next steps, I've put these forward as phase 2 project ideas basically to the project team.

kbarber replied:
So yeah, I've already asked about putting this point in some sort of errata/known issues for the release notes fwiw, most probably as a leading ticket as well people can vote/watch. Mainly, so its less surprising/documented somewhere.

Iristyle replied:
👍 sounds like you have a staged plan of attack that's reasonable.

I'll see if I can come up with situations where the addition of `idempotent` to a property could break module authors - at the very least it might be worth a quick grep through the modules for collisions (seems unlikely to dig anything up).

Iristyle replied:
FYI - I just threw `:idempotent => false` on some properties in existing types on the existing codebase, and ran the spec suite (just to see if anything would break).  I hoping this serves as a proxy for newer modules opting to use `:idempotent` running on an older Puppet where that concept doesn't exist.

````
diff --git a/lib/puppet/type/exec.rb b/lib/puppet/type/exec.rb
index f2526ba..11673e5 100644
--- a/lib/puppet/type/exec.rb
+++ b/lib/puppet/type/exec.rb
@@ -64,7 +64,7 @@ module Puppet
       @checks.keys
     end

-    newproperty(:returns, :array_matching => :all, :event => :executed_command) do |property|
+    newproperty(:returns, :array_matching => :all, :event => :executed_command, :idempotent => false) do |property|
       include Puppet::Util::Execution
       munge do |value|
         value.to_s
diff --git a/lib/puppet/type/file/content.rb b/lib/puppet/type/file/content.rb
index 614d9d7..9c98a58 100644
--- a/lib/puppet/type/file/content.rb
+++ b/lib/puppet/type/file/content.rb
@@ -6,7 +6,7 @@ require 'puppet/util/checksums'
 require 'puppet/type/file/data_sync.rb'

 module Puppet
-  Puppet::Type.type(:file).newproperty(:content) do
+  Puppet::Type.type(:file).newproperty(:content, :idempotent => false) do
     include Puppet::Util::Checksums
     include Puppet::DataSync

diff --git a/lib/puppet/type/package.rb b/lib/puppet/type/package.rb
index cab3c67..2b3e159 100644
--- a/lib/puppet/type/package.rb
+++ b/lib/puppet/type/package.rb
@@ -296,7 +296,7 @@ module Puppet
       [ [ /(.*)/m, [ [:name] ] ] ]
     end

-    newproperty(:package_settings, :required_features=>:package_settings) do
+    newproperty(:package_settings, :required_features=>:package_settings, :idempotent => false) do
       desc "Settings that can change the contents or configuration of a package.

         The formatting and effects of package_settings are provider-specific; any
diff --git a/lib/puppet/type/scheduled_task.rb b/lib/puppet/type/scheduled_task.rb
index 97215a2..1d6eb96 100644
--- a/lib/puppet/type/scheduled_task.rb
+++ b/lib/puppet/type/scheduled_task.rb
@@ -54,7 +54,7 @@ Puppet::Type.newtype(:scheduled_task) do
       should be specified as a space-separated string."
   end

-  newproperty(:user) do
+  newproperty(:user, :idempotent => false) do
     desc "The user to run the scheduled task as.  Please note that not
       all security configurations will allow running a scheduled task
       as 'SYSTEM', and saving the scheduled task under these
diff --git a/lib/puppet/type/service.rb b/lib/puppet/type/service.rb
index 1f8d266..8a5508c 100644
--- a/lib/puppet/type/service.rb
+++ b/lib/puppet/type/service.rb
@@ -122,7 +122,7 @@ module Puppet
       end
     end

-    newproperty(:flags, :required_features => :flaggable) do
+    newproperty(:flags, :required_features => :flaggable, :idempotent => false) do
       desc "Specify a string of flags to pass to the startup script."
     end

diff --git a/lib/puppet/type/user.rb b/lib/puppet/type/user.rb
index 3c1156e..78c9f36 100644
--- a/lib/puppet/type/user.rb
+++ b/lib/puppet/type/user.rb
@@ -98,7 +98,7 @@ module Puppet
       end
     end

-    newproperty(:home) do
+    newproperty(:home, :idempotent => false) do
       desc "The home directory of the user.  The directory must be created
         separately and is not currently checked for existence."
     end
```

All the specs passed on Windows:

```
Finished in 11 minutes 58 seconds (files took 15.14 seconds to load)
21219 examples, 0 failures, 62 pending
```

kbarber replied:
Cool.

kylog replied:
Fwiw, I'm 👍 on the use of `:idempotent` here as opposed to the whitelist approach. And thanks for that experiment @Iristyle - it seems like a reasonable proxy.

Iristyle commented on lib/puppet/transaction/resource_harness.rb:
```diff
@@ -92,6 +93,25 @@ def perform_changes(resource, context)
+  # application.
+  # @param [Puppet::Type] resource resource whose values we are to persist.
+  # @param [ResourceApplicationContent] context the application context to operate on.
+  def persist_system_values(resource, context)
```
I assume this will persist the actual file contents in a situation like this?

```puppet
file { '/tmp/foo': content => 'really_long_string', }
```

Iristyle replied:
Also - with respect to files... if the upstream source is a `puppet://` style URI -- and the backing contents of that file change, is that classified as a `corrective_change` or is it expected?

kbarber replied:
No, it stores the "shortened" version, depending on the provider property this can be different, for file its the md5.

> Also - with respect to files... if the upstream source is a puppet:// style URI -- and the backing contents of that file change, is that classified as a corrective_change or is it expected?

So this all translates to content property actions. In this case its much like setting the content field, so it won't be a corrective_change, its intended.

Iristyle replied:
Ah, thanks for clarifying.

kbarber replied:
I was worried about space too. Our original plan was to do this all in PuppetDB, but we only get the catalog which contains the full string, so it would have been hard to get the md5 instead etc. Keeping it in the agent, we benefit from having these short-forms in hand, and its native comparison capability also, which makes the solution accurate (well as accurate as insync?).

Iristyle commented on lib/puppet/transaction/event.rb:
```diff
@@ -58,11 +60,13 @@ def to_data_hash
   end
 
   def property=(prop)
+    @property_instance = prop
```
There might have been a reason that the property object itself was never cached previously.  Not sure if @joshcooper might know why.

kbarber replied:
Be interesting to know, I was careful to have this avoid the YAML serialization at least. Its very easy to accidentally drop crap into the report for example without realizing it :-).

kylog replied:
I spelunked back to 2009 on this (https://github.com/puppetlabs/puppet/commit/977595bd if you're curious) and I didn't find a reason why the property object itself wasn't cached. My *guess* is that it was following a pattern of "normalize things to strings" and there just hadn't been a reason before. But dunno.

That said, I'm not coming up with a reason adding this would cause problems.

Iristyle commented on lib/puppet/transaction/persistence.rb:
```diff
@@ -0,0 +1,81 @@
+# transaction runs for the purposes of information inference (such
+# as calculating corrective_change).
+# @api private
+class Puppet::Transaction::Persistence
```
The last run report can't be extended to meet your needs I presume (I'm not entirely familiar with the difference b/w this new YAML file and what we already have)

kbarber replied:
We went down that road also, kylo and eric0 had concerns about extending the report format this way, when we had no intention of using it directly (ie. PuppetDB wasn't going to consume these system values, nor was it ready for that). Since its a rather public API, we decided it best to avoid that kind of thinking for later.

Iristyle replied:
Ah, so you've already been chatting with those guys on this - great.  That reasoning all makes sense.

Sorry I'm a bit late to the party then - carry on!

kbarber replied:
Its cool, its nice to know I wasn't the only one to consider this path :-). But yeah, devil in the details.

joshcooper commented:
I'm joining this conversation in the middle, apologies if I'm asking questions that have already been answered elsewhere.

1. Are we're reinventing `audit`functionality, keeping track of old values so that we can compare against current values and infer if it was changed by something other than puppet in-between? Guess I'm a bit surprised because we've been talking about deprecating audit for a long time: https://tickets.puppetlabs.com/browse/PUP-893.
1. Does it make sense to mark the entire resource as idempotent or not, e.g. `notify`, `exec`, instead of individual properties?
1. Would it make sense to create an `idempotent` metaparam like `audit` instead of modifying the type definitions? That way users (not module authors) can opt-in/out easily? Or a combination of both (define default behavior in the type and allow opt-in/out in manifest)?
1. `Puppet::Transaction::Persistence` behaves just like `Puppet::Util::Storage`, plus adds getter/setter methods for the data you care about. Why not use `Storage` to handle the persistence and have your class provide methods for manipulating the "subtree" of yaml data you care about?
1. Since the report schema is changing, we should bump the report version/format
1. I didn't follow the comments in https://github.com/puppetlabs/puppet/pull/5139/files#diff-93ff56b1885dd711e694649d0cad0b77R374. Why do we need to create a temp property? Does the `dup` correctly handle structured values, I'm thinking of the Windows ACL module specifically.

adrienthebo commented:
I can answer some of these questions.

1. We're partially reimplementing audit, but we're doing so because audit caches the historical state before Puppet makes changes; we need the historical state cached after Puppet makes changes. (Right, @kbarber?)
2. I don't think so; exec can be idempotent when configured correctly. Conversely, service resources have had a history of becoming non-idempotent (specifically the `enable` property when using systemd with a legacy init script), but the resource itself should be idempotent.
3. Maybe? Allowing resources to be easily marked as non-idempotent feels like we're enabling/supporting misbehaving types on an ad-hoc basis, whereas with notify the intended behavior is non-idempotent.
4. `P::T::Persistence` is based on `P::Util::Storage`, but the main difference is that `Storage` doesn't discard old information from resources that were removed and will bloat forever, while `Persistence` throws away information about resources that were removed.
5. We've already bumped the report format version for the 4.6.0 release for a couple of other features, so we don't need to bump it again.
6. We're using `Puppet::Property#insync?` to detect drift from a recorded historical value. Because `#insync` relies on `@should` we can't just call that method with a new `@should` value; we have to replace the value entirely. This issue is the reason why I wanted to bring more people into this conversation; I can't say for certain that this will be safe. It might be, but it also might break all sorts of things depending on what people have done with properties. My understanding is that `#dup` just copies all of the instance variables to a new object, and will call `#initialize_copy` if it's defined, but that should be the scope of the operation.

kbarber commented:
1. See Adrien's response. Audit doesn't store the right thing, but it was so close. I looked into the audit file, and almost was convinced, but it has issues and others talked me out of using it also: a) it had a hard API to change file-wise (values assigned directly to keys, I couldn't just add my new value) b) its semantics are more around making small transactional changes to a file c) it has no cleanup, just yesterday I had a guy on IRC complaining, and my advice was to stop using it basically d) everyone in the team I spoke to had issues with it, so it felt like beating a dead horse to improve it. So basically, yes, its close but not quite to audit, however if you wanted to use this instead of audit its possible, we architecturally have left that an open door for future conversations. In a way though, philosophically, this entire functionality kind of does replace some of audit's/inspects goals with a slightly more modern flair perhaps, so if/when that conversation happens about replacing audit, we might want to take this into account. Great question btw.
2. You make an interesting point re: property v resource. Let me ponder it.
3. Nice idea. Its a maybe/yes for me too. We can also add that feature too later right?
4. See Adrien's response.
5. Ditto.
6. I added unit tests to property_spec.rb to test this method against the same stuff insync? is tested with. There are some arrays in there, but we can go deeper structure if you're concerned no problemo. The reason I dup, is because I can't get setting a should to a value working without a dup, it seemed like a reasonable alternative but everyone dislikes it. I've been trying to figure out why just forcing should to something, do the insync? then restoring should doesn't work, but it makes no sense. Perhaps if this is a critical thing (and please, say so) I could reproduce it and seek help from someone. At the very least, we moved this weirdness into its own place to contain it.

@joshcooper cheers :-).

kbarber commented:
2. (after pondering some more) it should be the per property at the very least. In pondering this, I could see puppetlabs-firewall (which is chock full of properties) having idempotency issues that are specific to say 1 property, but the others are fine (happened a lot due to its shoddy parser, sigh). However, I could see convenience added to an entire resource as sugar at any point in the future, if it became worth it.

adrienthebo commented:
My understanding is that we're under a time crunch for this feature and that it's blocking a lot of other work, so I want to push this PR forward. While there are some outstanding concerns about the use of `Puppet::Property#dup`, I don't know of any specific case where this is guaranteed to break. @joshcooper @Iristyle if either of you know of a case where this could cause issues, we need to know very soon.

I see that [puppetlabs-acl](https://github.com/puppetlabs/puppetlabs-acl/blob/master/lib/puppet/type/acl/ace.rb#L24-L44) does have some nontrivial `#initialize` behavior, but since `#dup` should only copy internal variable state and shouldn't initialize the object I think we're okay. For the other properties I've seen with some custom initialization logic it's nothing too dangerous, so all in all I think we can get away with this.

While it would be nice to have a feature flag to allow this feature to be turned off if needed I don't think we have the time to build that in, so it's kind of an all or nothing proposition.

In general, unless anybody has a firm reason to reject this PR or halt it for a specific reason, I think we should merge this. We should definitely revisit this after the release and address some of the outstanding concerns, but for the sake of getting this work in so that other work can complete I'm :+1: for merge. I'm currently at a conference so I don't feel comfortable merging this myself, but @branan if you could do this or get someone else to merge it and shepard it for CI, I'll buy you lunch when I get back.

/cc @kbarber @scottyw @ploubser

branan commented on api/schemas/report.json:
```diff
@@ -160,6 +160,11 @@
+        },
+
+        "corrective_change": {
+            "description": "True if this report suffered an unexpected change, and Puppet corrected it, or was intending to correct it if the resource or entire run was noop.",
```
How about "True if a change or noop event was caused by an unexpected change to the system between Puppet runs"?

kbarber replied:
Fixed.

branan commented on lib/puppet/property.rb:
```diff
@@ -102,6 +102,22 @@ def array_matching=(value)
+    # generally). This is used to avoid marking the property as a
+    # corrective_change when there is known idempotency issues with the property
+    # rendering a corrective_change flag as useless.
+    # @return [Boolean] true if the property is marked as non-idempotent
```
"Returns true if the property is non-idempotent" sounds backwards. Is that just a docs issue?

kbarber replied:
Fixed.

branan commented on lib/puppet/property.rb:
```diff
@@ -346,6 +362,29 @@ def insync?(is)
+  # that requires two older values to be compared with the properties potentially
+  # custom insync? code.
+  #
+  # @param [Object] orig_should the original should value
```
The phrasing of "original" here definitely shows its heritage from corrective_change. These should probably just be "should" and "is", and the documentation made more generic. This can be a post-merge cleanup, IMO

kbarber replied:
Fixed.

kylog commented on lib/puppet/transaction/event.rb:
```diff
@@ -94,6 +98,25 @@ def to_yaml_properties
+      @corrective_change = false
+    end
+
+    return @corrective_change
```
I think the `return` can be dropped. Actually I think the line can be dropped.

kbarber replied:
fixed.

joshcooper commented:
What happens when the agent changes environment? Do we still report corrective changes between the previous run (that happened in a different environment)? Does that make sense?

Related, do we persist previous system values during a noop run? If so that seems bad, because it will lead to false alerts. I don't think it blocks this PR, but seems like something we should fix.

kylog commented:
@kbarber total nit, but can you amend the commit message to be in the imperative, e.g. "Add corrective_change field to report and report events" or some such. Buried in https://github.com/puppetlabs/puppet/blob/master/CONTRIBUTING.md is the request to make summaries imperative.

kbarber commented:
@joshcooper if puppet changes environment, its fine. We compare the last set values to the value we get from a `#retrieve`, we don't care what puppet is changing things to, so I'm not sure how an environment change could affect things against this patch.

During a noop run, we persist the old stored system value, and throw a correct_change. Thus we still get a corrective_change for each noop run, until its finally corrected and the new set value is set in the system_value.

kbarber commented:
@kylog done.

kylog commented:
I'm 👍 on this PR as-is. There are several good follow-ups in the extensive commenting above; it'd be nice to pull actionable items out into the appropriate places (probably tickets or docs).

branan commented:
I agree with Kylo. Let's get actionable stuff someplace where we can track it, and this is fine to merge at this point

kbarber commented:
I'm sorry, I can't tell what's important to people or not any more. What do people want me to document or raise tickets on?

kylog commented:
Oh good, travis sprung back (it seemed to have lost track of the PR for a bit). Once travis and appveyor flip green, I'll merge.

joshcooper commented:
> During a noop run, we persist the old stored system value, and throw a correct_change. Thus we still get a corrective_change for each noop run:

Sorry, what I mean is, it seems incorrect to call [`persistence.save`](https://github.com/puppetlabs/puppet/pull/5139/files#diff-129f1c7cf7c3e8df7a5586dde253d845R148) during a noop run as that will alter the stored system values. So when the agent next runs (not in noop), it will be confused about what the previous system value was, and won't correctly flag the corrective_change. I'm fine to merge, we can discuss separately.

kylog commented:
@kbarber for starters, I'm inclined to send you a bottle of your beverage of choice :) I know this PR was a gauntlet. Thanks for all your work on this.

And re action items, after lunch I'll take a pass through the PR to cull out possible follow-up items and drop another comment in the PR.

kbarber commented:
@joshcooper so the short answer here is, it just works? Meaning, it doesn't get confused ... but perhaps I'm not explaining the implementation well. It doesn't care about the overall noop run state true, however it behaves differently when it sees that a event was a non-success however: https://github.com/puppetlabs/puppet/pull/5139/files#diff-466a7aa2dfd3724f006c1984f9ce87acR270 ... I agree there is probably some efficiency to be found here potentially, but with the uncertainty of perhaps some resources becoming noop => false even when the entire runs is noop (I can't recall how, but I've heard this is possible) I'm inclined to think I took the safer approach for now. Oh, and this also works for failed, skipped etc.

kylog commented:
Okay, @kbarber, I re-read the comment thread for follow-ups and I mostly concluded that this change needs release notes. In terms of mechanics, I'd think you could just use the Release Notes Summary field in https://tickets.puppetlabs.com/browse/PUP-6295, though maybe there ends up being enough content for a pre-docs PR.

In terms of content: offhand, in addition to explaining the general idea of the `corrective_change` fields, I'd think the release notes would cover use of `:idempotent` and interaction with the new Sensitive type. Also, in https://github.com/puppetlabs/puppet/pull/5139#discussion_r72809223, Ken, you mentioned release notes errata (and a leading ticket) re conditions where this *doesn't* work.

/cc @jtappa as a heads-up that a somewhat substantive release note will be inbound for puppet 4.6.

Reviewers of this PR: if I missed some follow-ups, please chime in!
